driver-core/src/main/com/mongodb/internal/connection/BaseCluster.java

```diff
@@ -280,19 +263,20 @@ private void updatePhase() {
         withLock(() -> phase.getAndSet(new CountDownLatch(1)).countDown());
     }
 
-    private long getMaxWaitTimeNanos() {
-        if (settings.getServerSelectionTimeout(NANOSECONDS) < 0) {
-            return Long.MAX_VALUE;
-        }
-        return settings.getServerSelectionTimeout(NANOSECONDS);
+    private Timeout startServerSelectionTimeout(final StartTime startTime) {
+        long ms = settings.getServerSelectionTimeout(MILLISECONDS);
+        return startTime.timeoutAfterOrInfiniteIfNegative(ms, MILLISECONDS);
     }
 
-    private long getMinWaitTimeNanos() {
-        return serverFactory.getSettings().getMinHeartbeatFrequency(NANOSECONDS);
+    private Timeout startMinWaitHeartbeatTimeout() {
+        long minHeartbeatFrequency = serverFactory.getSettings().getMinHeartbeatFrequency(NANOSECONDS);
+        minHeartbeatFrequency = Math.max(0, minHeartbeatFrequency);
```

Comment:
Nice - I dont think there is a negative check or minimum value for this.

Reply:
Yeah, anything negative seems like it is effectively 0. I'm not sure what good a value like 0 or 1 would be, but there's nothing preventing these.

I added a TODO that I'll mark in the ultimate CSOT PR to check that we don't accept negatives for timeouts.